### PR TITLE
 Contact Journal Extension - Store RAT and PCR Tests results (EXPOSUREAPP-6041)

### PR DESCRIPTION
**Pull request by jurajkusnier**
- Test result is added to a new database and later will be displayed in the contact journal 
 
### Testing
- Scan PCR/RAT test
- See test results
- If the test is positive or negative new record should be added to DB
- Check Test Menu > Corona Tests > Contact diary tests if the test is there

### CODE REVIEW

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/coronatestjournal/TestJournalRepository.kt [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package de.rki.coronawarnapp.coronatestjournal
```
Comment (BMItter) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/coronatestjournal/TestJournalRepository.kt:1]: package coronatestjournal.. (all relevant files)
everything is only needed in Contactjournal package.
No need to create a new separate package

File: Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/test/coronatest/ui/CoronaTestTestFragment.kt [hunk 2]
```diff
@@ -65,26 +65,30 @@ class CoronaTestTestFragment : Fragment(R.layout.fragment_test_coronatest), Auto
             qrcodeScanViewfinder.setCameraPreview(binding.qrcodeScanPreview)
         }
 
-        vm.pcrtState.observe2(this) {
+        viewModel.pcrtState.observe2(this) {
```
Comment (BMItter) [inline Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/test/coronatest/ui/CoronaTestTestFragment.kt:68]: vm -> viewModel
how is the naming in other classes?

**Comment by BMItter** [comment]
@jurajkusnier 
Looking good overall 👌 
I would consider not to create a new DB for the TestResults, because the ContactJournal already has an own DB.
Just create a new Table for your needs..

File: Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/test/coronatest/ui/CoronaTestTestFragment.kt (hunk 2 above)
Comment (jurajkusnier) [reply Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/test/coronatest/ui/CoronaTestTestFragment.kt:68]: usually, we call it `viewModel` in a new code, so I used this opportunity for a small refactoring

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/coronatestjournal/storage/TestJournalStorage.kt [hunk 3]
```diff
@@ -0,0 +1,47 @@
+package de.rki.coronawarnapp.coronatestjournal.storage
+
+import de.rki.coronawarnapp.coronatest.qrcode.CoronaTestGUID
+import de.rki.coronawarnapp.coronatest.type.CoronaTest
+import de.rki.coronawarnapp.coronatest.type.rapidantigen.RACoronaTest
+import javax.inject.Inject
+
+class TestJournalStorage @Inject constructor(
```
[the remark this answers is not in the document]
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/coronatestjournal/storage/TestJournalStorage.kt]: updated

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/coronatestjournal/TestJournalRepository.kt (hunk 1 above)
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/coronatestjournal/TestJournalRepository.kt]: updated

**Comment by jurajkusnier** [comment]
> @jurajkusnier
> Looking good overall 👌
> I would consider not to create a new DB for the TestResults, because the ContactJournal already has an own DB.
> Just create a new Table for your needs..

Thanks, yes, it makes a perfect sense. I've updated the code

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/storage/internal/converters/ContactDiaryRoomConverters.kt [hunk 4]
```diff
@@ -25,4 +26,18 @@ class ContactDiaryRoomConverters {
     fun fromJodaDuration(duration: Duration?): Long? {
         return duration?.millis
     }
+
+    @TypeConverter
+    fun toTestType(value: String?): ContactDiaryTestEntity.TestType? =
+        value?.let { ContactDiaryTestEntity.TestType.values().single { it.raw == value } }
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/storage/internal/converters/ContactDiaryRoomConverters.kt:32]: `singleOrNull` ? single throws an exception if not found

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/storage/internal/converters/ContactDiaryRoomConverters.kt [hunk 5]
```diff
@@ -25,4 +26,18 @@ class ContactDiaryRoomConverters {
...
+
+    @TypeConverter
+    fun fromTestType(type: ContactDiaryTestEntity.TestType?): String? = type?.raw
+
+    @TypeConverter
+    fun toTestResult(value: String?): ContactDiaryTestEntity.TestResult? =
+        value?.let { ContactDiaryTestEntity.TestResult.values().single { it.raw == value } }
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/storage/internal/converters/ContactDiaryRoomConverters.kt:39]: Here too?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/DataReset.kt [hunk 6]
```diff
@@ -66,7 +66,7 @@ class DataReset @Inject constructor(
     private val ratProfileSettings: RATProfileSettings,
     private val valueSetsRepository: ValueSetsRepository,
     private val vaccinationPreferences: VaccinationPreferences,
-    private val vaccinationRepository: VaccinationRepository,
+    private val vaccinationRepository: VaccinationRepository
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/DataReset.kt:69]: why? 
I think we will add more repos in the future here

Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/DataReset.kt]: it wasn't intentional.
reverted

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/storage/entity/ContactDiaryTestEntity.kt [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        ContactDiaryTestEntity(
+            id = key,
+            testType = if (type == CoronaTest.Type.PCR)
+                ContactDiaryTestEntity.TestType.PCR
+            else
+                ContactDiaryTestEntity.TestType.ANTIGEN,
+            result = if (isPositive)
+                ContactDiaryTestEntity.TestResult.POSITIVE
+            else
+                ContactDiaryTestEntity.TestResult.NEGATIVE,
```
[the remark this answers is not in the document]
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/storage/entity/ContactDiaryTestEntity.kt]: yes, could be more readable. updated

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/storage/internal/converters/ContactDiaryRoomConverters.kt (hunk 4 above)
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/storage/internal/converters/ContactDiaryRoomConverters.kt]: updated

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/storage/internal/converters/ContactDiaryRoomConverters.kt (hunk 5 above)
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/storage/internal/converters/ContactDiaryRoomConverters.kt]: updated

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/storage/entity/ContactDiaryTestEntity.kt [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+import androidx.room.ColumnInfo
+import androidx.room.Entity
+import androidx.room.PrimaryKey
+import com.google.gson.annotations.SerializedName
+import de.rki.coronawarnapp.coronatest.qrcode.CoronaTestGUID
+import de.rki.coronawarnapp.coronatest.type.CoronaTest
+import de.rki.coronawarnapp.coronatest.type.rapidantigen.RACoronaTest
+import org.joda.time.Instant
+
+@Entity(tableName = "tests")
```
[the remark this answers is not in the document]
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/storage/entity/ContactDiaryTestEntity.kt]: updated to `corona_tests`

**Comment by BMItter** [comment]
@jurajkusnier thx for the adjustments.
Please consider at last to add some tests for migration & table and/or adjust the  "ContactDiaryDataBaseMigrationTest" & "ContactDiaryDatabaseTest".
Also testing the converter would make some sense (but low prio. optional)

**Review by mtwalli — DISMISSED** [review]
Lgtm ,please consider Bernd points ?

**Review by mtwalli — APPROVED** [review]
👍🏼

**Review by BMItr — APPROVED** [review]
thx for the final adjustments.
lgtm! 👌

**Pull request merged**